drivers/adc/CMakeLists.txt

```diff
@@ -1,8 +1,10 @@
-zephyr_sources_ifdef(CONFIG_ADC_DW            adc_dw.c)
-zephyr_sources_ifdef(CONFIG_ADC_MCUX_ADC16 	  adc_mcux_adc16.c)
-zephyr_sources_ifdef(CONFIG_ADC_QMSI   		  adc_qmsi.c)
-zephyr_sources_ifdef(CONFIG_ADC_QMSI_SS		  adc_qmsi_ss.c)
-zephyr_sources_ifdef(CONFIG_ADC_SAM_AFEC      adc_sam_afec.c)
-zephyr_sources_ifdef(CONFIG_ADC_TI_ADC108S102 adc_ti_adc108s102.c)
+zephyr_library_ifdef(CONFIG_ADC_HAS_DRIVER)
```

Comment:
The need for _ifdef(ADC_HAS_DRIVER) is not clear to me.

I would think 

zephyr_library()

would be enough since drivers/CMakeLists.txt

is already doing:

add_subdirectory_if_kconfig(adc)

Reply:
"During the build someone would enable the CONFIG_ADC but before enabling this they don’t check if the given soc has a driver associated to it."

I still don't understand.

CONFIG_ADC_HAS_DRIVER means that the SoC supports the ADC driver.
CONFIG_ADC means that the user wishes to enable the ADC driver.

This CMakeLists.txt code is expressing that an ADC library should be enabled when an SoC supports ADC. But that is not correct, libraries should be enabled when the user has enabled
it. And enabling it should depend on the SoC supporting the driver. The dependency of having an enabled driver, and the driver being supported should be expressed in Kconfig.

So it should be something like this:

zephyr_library_ifdef(CONFIG_ADC)

config ADC
  depends on CONFIG_ADC_HAS_DRIVER

Reply:
In some test cases or samples they enable the driver but for the soc they write custom driver code or dummy code and they continue to use the Kconfig options available when then enable the driver. 
By doing this(as in the PR) I say, you can enable the driver but the soc specific code for the driver if not present in the drivers/{name} folder I don't compile anything inside it. But you can access any of the Kconfigs which get enabled when by saying CONFIG_ADC=y.
 
I think the method I used works without breaking any samples/testcases . This way its backward compatible. I feel more secure doing it this way(as in the PR).

Reply:
"In some test cases or samples they enable the driver but for the soc they write custom driver code or dummy code and they continue to use the Kconfig options available when then enable the driver."

What you are describing is an invalid use of the configuration system. We cannot write (invalid) build scripts to make invalid configurations build. Instead, the invalid use must be rectified.

One problem with the scheme that you are proposing is that we can no longer have the configuration system enforce valid configurations. E.g. if there is no dependency
 of enabling ADC on ADC being supported, we can not prevent users from creating invalid configurations. Enforcing valid configurations is one of the primary objectives of the 
configuration system so it is not something we can drop.

It is okay to break samples/tests when they rely on invalid behaviour.

Reply:
@SebastianBoe I understand the concern here. I tried doing this for FLASH and ADC. there are too many dependencies and it is not straight forward and is very tedious. Also this would not be in the scope of this PR. 
So the best thing I can see is either we continue with this approach or we do not include _supervisor_ macro for drivers. I am inclined to not include the macro.
If and when we need the optimization for the drivers, we can take this up then.
(@andrewboie  what do you think?)

Reply:
Fair enough, you have a point that a failure in one subsystem should not halt progress in another.

But if we approve this PR then this means we are in agreement that:

We are writing Kconfig in a way that does not allow enforcement of valid configurations.

We do not know how to write Kconfig in a way that allows enforcement of valid configurations,
and therefore must continue to write invalid Kconfig until we can give guidelines that can be enforced through code-review.

Reply:
this needs more thought, I dislike this change adding all of those Kconfig dependencies and hidden variables. Will sync with @AdithyaBaglody to understand the issue better and possibly come up with a better solution.